vmm/src/api/http/http_endpoint.rs

```diff
@@ -50,6 +52,162 @@ use crate::config::RestoreConfig;
 use crate::cpu::Error as CpuError;
 use crate::vm::Error as VmError;
 
+mod helper {
+    use std::fs::File;
+    use std::os::fd::IntoRawFd;
+
+    use crate::api::http::HttpError;
+
+    /// Abstraction over configuration types received via the HTTP API that
+    /// logically hold FDs. Helper for [`apply_new_fds_to_cfg`].
+    ///
+    /// Example: virtio-net device with externally opened FDs for the Tap device
+    pub trait ConfigWithFDs {
+        /// Returns the ID of the device.
+        ///
+        /// Used for logging.
+        fn id(&self) -> Option<&str>;
+
+        /// Returns how many FDs this type wants to own.
+        ///
+        /// `None` instead `Some(0)` indicates that this type wants to take
+        /// ownership of all available FDs. This is the case for config options
+        /// that don't know (yet) how many FDs belong to them, such as on hot
+        /// device attach. In these cases, only one single config object is
+        /// handled by the API call.
+        fn expected_num_fds(&self) -> Option<usize>;
+
+        /// Returns any FDs provided in the HTTP body.
+        ///
+        /// They will always be invalid. Used for logging.
+        fn fds_from_http_body(&self) -> Option<&[i32]>;
+
+        /// Updates the internal fds field to a new value.
+        fn set_fds(&mut self, fds: Option<Vec<i32>>);
```

Comment:
I'm not sure what the doc comment adds here…

Reply:
Do you like the new comment more?

```
        /// Assigns the provided file descriptors (`fds`) to this configuration
        /// object.
        ///
        /// After calling this method, the configuration will behave as if it
        /// had originally been created with these FDs. Next, the configuration
        /// can be used to properly configure the corresponding device.
        ///
        /// # Arguments
        /// - `fds`: Either a non-empty Vector with corresponding FDs or `None`
        ///   indicating that no valid FDs were supplied.
```